### ISSUE TITLE
Strong deviations between Hyperopt and Backtest Results

### ISSUE THREAD

**Issue by Prossi79** [freqtrade/freqtrade#2008]
## Step 1: Have you search for this issue before posting it?

If you have discovered a bug in the bot, please [search our issue tracker](https://github.com/freqtrade/freqtrade/issues?q=is%3Aissue). 
If it hasn't been reported, please create a new issue.

## Step 2: Describe your environment

  * Python Version: 3.7.3
  * CCXT version: pip freeze | grep ccxt
  * Branch: Develop
  * Last Commit ID: d055dc0c6e0aa24297e5309eee7ea7e511bf9e51

  
## Step 3: Describe the problem:

Strong deviations between Hyperopt and Backtest Results, **number of trades and total profit**

### Steps to reproduce:

  1. strategy (MACD and CCI based) coded in Hyperopt to find the best parameters for the CCI buy / sell, stoploss and ROIs
  2. run Hyperopt
  3. put the paramters in the strategy and backtested over the same timeframe as hyperopt was running
  
### Observed Results:

Results should be similar, correct?

### Relevant code exceptions or logs:

hyperopt results

  ```
2019-07-09 07:01:39,815 - freqtrade.exchange.exchange - INFO - downloaded LINK/BTC with length 1.
2019-07-09 07:01:39,816 - freqtrade.misc - INFO - dumping json to "user_data/data/binance/LINK_BTC-5m.json"
2019-07-09 07:01:40,851 - freqtrade.data.converter - INFO - Missing data fillup for LINK/BTC: before: 71711 - after: 71999
2019-07-09 07:01:40,884 - freqtrade.optimize.hyperopt - INFO - Hyperopting with data from 2018-11-01T05:05:00+00:00 up to 2019-07-09T04:55:00+00:00 (249 days)..

2019-07-09 09:43:45,460 - freqtrade.optimize.hyperopt - INFO - Best result:
   604 trades. Avg profit  0.53%. Total profit  0.48081498 BTC ( 320.22Σ%). Avg duration 1292.9 mins.
with values:

{   'buy-cci-value': -12,
    'roi_p1': 0.014714316102334512,
    'roi_p2': 0.06416378274370114,
    'roi_p3': 0.12198411469836923,
    'roi_t1': 114,
    'roi_t2': 43,
    'roi_t3': 33,
    'sell-cci-value': 158,
    'stoploss': -0.4700606834872094}


  ```

Backtest results

  ```

2019-07-09 16:59:11,178 - freqtrade.optimize.backtesting - INFO - Backtesting with data from 2018-11-01T00:00:00+00:00 up to 2019-07-08T23:55:00+00:00 (249 days)..
2019-07-09 16:59:11,178 - freqtrade.optimize.backtesting - INFO - Running backtesting for Strategy MACDStrategy1
Result for strategy MACDStrategy1
========================================================= BACKTESTING REPORT ========================================================
| pair     |   buy count |   avg profit % |   cum profit % |   tot profit BTC |   tot profit % | avg duration    |   profit |   loss |
|:---------|------------:|---------------:|---------------:|-----------------:|---------------:|:----------------|---------:|-------:|
| LINK/BTC |          88 |           0.63 |          55.61 |       0.08349766 |          18.54 | 2 days, 6:31:00 |       33 |      1 |
| TOTAL    |          88 |           0.63 |          55.61 |       0.08349766 |          18.54 | 2 days, 6:31:00 |       33 |      1 |
========================================================= SELL REASON STATS =========================================================
| Sell Reason   |   Count |
|:--------------|--------:|
| roi           |      87 |
| force_sell    |       1 |
====================================================== LEFT OPEN TRADES REPORT ======================================================
| pair     |   buy count |   avg profit % |   cum profit % |   tot profit BTC |   tot profit % | avg duration   |   profit |   loss |
|:---------|------------:|---------------:|---------------:|-----------------:|---------------:|:---------------|---------:|-------:|
| LINK/BTC |           1 |          -5.67 |          -5.67 |      -0.00851834 |          -1.89 | 11:40:00       |        0 |      1 |
| TOTAL    |           1 |          -5.67 |          -5.67 |      -0.00851834 |          -1.89 | 11:40:00       |        0 |      1 |


  ```

**Comment by hroff-1902** [comment]
@xmatthias It seems I found the difference.
"position_staking": True for hyperopt and False for backtesting in the args to `backtest()`.

Did not find the reason yet...

**Comment by hroff-1902** [comment]
And the results of hyperopt are now realistic, hooray! 

Previously the results were simply impossible (many long trades while there is only one pair in the whitelist set in the config). position_stacking: True describes that as well.

**Comment by xmatthias** [comment]
I'll be honest, i am not convinced of your fix:

The documentation [here](https://www.freqtrade.io/en/latest/hyperopt/#validate-backtest-result) explains why this is the case (disabling position-stacking has a performance hit - which is not that clear in backtesting, but becomes very relevant if you run backtesting 1000nds of times (as with hyperopt) - then 1-2 more seconds per run (for half a year / 20ish pairs) does make a ton of a difference.

Best test this with backtesting with /without the `--enable-position-stacking` flag and a "longer" timeframe (at least 1-2 months, more is better)

Changing this default will have hyperopt take multiple times as long ... which may not be the desired thing.

**Comment by hroff-1902** [comment]
Can you explain me the purpose of the position_stacking option at all?

In my opinion, all results given by backtesting or hyperopt with this setting set to True are falsy since main freqtrade cannot trade stacking the positions... Those results, in my opinion, are simply useless...

**Comment by hroff-1902** [comment]
backtesting (and, thus, hyperopt) is quicker with --eps just because the time of `backtest()` depend on time of trades in the backtesting evalutions in very close degree. While it should not, as I can see. So by setting eps to True __by default__ you try to speed up the hyperopt but in this case it produces useless data since main hyperopt cannot be run in this position stacking mode...

**Comment by hroff-1902** [comment]
A user, looking at the hyperopt results, produced __by default__, i.e. with eps set to True, can decide that she found a Holy Grail since with eps=True the profit (depending on the strategy) may produce profits 5-10 times bugger than with eps=False. And then, using the found strategy params, she is frustrating of real results. It's not only about the deviation between backtesting and hyperopt...

**Comment by hroff-1902** [comment]
Aha, I found in the issues that these options were previously combined in one with the name `--realistic-simulation`. The default for hyperopt thus really produces unrealistic data... 😄

**Comment by hroff-1902** [comment]
> For hyperopt I think it should default to true, it rarely makes sense running hyperopting for "just the first found trades".

See https://github.com/freqtrade/freqtrade/pull/1037/commits/e17618407bc3abb75eaed957b2faf350e45088a9#r203605324. Actually I did not get what "just the first found trades" can mean.

As from reviewing the code and from ongoing experiments, I see that --eps only slows down calculation of `backtest()`.

**Comment by hroff-1902** [comment]
Also found this:
> change realistic simulation to default to false and allow enabling it from both hyperopt and backtesting with --realistic-simulation

See https://github.com/freqtrade/freqtrade/pull/518#issue-169704277. Seems in Feb 2018 --realistic-simulation was default for __both__ hyperopt AND backtesting. When it got changed, hmm?

**Comment by hroff-1902** [comment]
I found this:
> change realistic simulation to default to false and allow enabling it from both hyperopt and backtesting with --realistic-simulation

See https://github.com/freqtrade/freqtrade/pull/518#issue-169704277. No explaination why default for --realistic was made False. 

@xmatthias maybe you can find some more traces of the decisions made regarding these options... 

I'm just trying to understand why --realistic option was introduced (why not simply implement this realistic approach in backtesting/hyperopt and remove unrealistic approach?) What was the reason?

**Comment by xmatthias** [comment]
github has an edit function ... this makes it hard to follow you tbh.

To at least answer partially (i think you found most answers to the above yourself?):
these 2 options (split from --realistic) allow you to test your buy/sell signals. During strategy development, i may not  care about max-open-trades, nor about position stacking.
The final strategy needs to care about this - but during development, immagine the following situation:

I have one pair in my pairlist
my strategy produces 5 buy signals - 4 of them are good - one is bad. 
Unfortunately, that bad one happens to be the first one, which holds throughout the whole backtesting period - so this "hides" the other 4, actually good, buy signals. 
I now need a way to filter this signal for the real strategy - however i would have toss this strategy away if i don't know about the 4 valid trades.

`--dnmp` makes even more sense, immagine the following:

20 pairs in pairlist, max-trades to 3 (the default).
the first 3 pairs it buys are "bad" buys - which hold throughout the whole backtest period. This hides 25 other trades, which would happen on the other 17 pairs, so disabling this during backtesting can make sense.

Remember that backtesting is set to "--realistic" by default now.
This was not done for hyperopt - however back when we split `--realistic` to the two new options, there was a huge performance hit when running `--realistic`. 

I'll need to do tests if that's still the case - especially with more (longer) data.

**Comment by hroff-1902** [comment]
> To at least answer partially (i think you found most answers to the above yourself?):

I found some historical background but did not find any reasonable motives why it is designed in this way. Your explaination above clarifies it, thanks. That's what I wanted to know.

**Comment by hroff-1902** [comment]
> The final strategy needs to care about this - but during development, immagine the following situation:
> 
> I have one pair in my pairlist
> my strategy produces 5 buy signals - 4 of them are good - one is bad.
> Unfortunately, that bad one happens to be the first one, which holds throughout the whole backtesting period - so this "hides" the other 4, actually good, buy signals.
> I now need a way to filter this signal for the real strategy - however i would have toss this strategy away if i don't know about the 4 valid trades.
> 
> `--dnmp` makes even more sense, immagine the following:
> 
> 20 pairs in pairlist, max-trades to 3 (the default).
> the first 3 pairs it buys are "bad" buys - which hold throughout the whole backtest period. This hides 25 other trades, which would happen on the other 17 pairs, so disabling this during backtesting can make sense.
> 
> Remember that backtesting is set to "--realistic" by default now.
> This was not done for hyperopt - however back when we split `--realistic` to the two new options, there was a huge performance hit when running `--realistic`.
> 

Okay, and I even can agree with most of it. I got the points.

But strategies do not run in vacuum. Imagine we have a strategy that is _designed_ (or assumed) to work in the freqtrade frames. For example, the MACDStrategy from freqtrade-strategies. It uses comparison between macd and macdsignal (not crosses) and will produce the long raws of buy and sell signals. It's pretty simple, users could often write such simple strategies, especially when they take the examples from freqtrade-strategies, and it's simple just because it uses the knowledge that it will run in the frames of freqtrade implementation where only one open order for a pair is possible in every moment of time. In the same time, it's a correct strategy for freqtrade which can even be profitable.

When you run hyperopt build for this strategy with default settings, it will produce many trades, for any candle in a raw of candles with buy signal. The results are unreaslistic. And even if freqtrade would have the position stacking implemented, these results will still be unrealistic because noone would accept a hundred of orders open for one pair in reality.

A modification of that strategy, with crossed_above/crossed_below used instead of comparison of values of two indicators behaves better with default hyperopt settings. But the problem here is that there are lot's of similar strategies in freqtrade-strategies and they are assumed to be examples that can inspire novice users and can be used as templates.

At least, all this should be covered in details in the docs. With examples and explainations on how to treat the hyperopt and backtesting results in different modes.

I was going to close #2009, but did not make it because of two points:

1) Inconsistency in the options. If the default mode for hyperopt is eps=True, there should be a cli option to disable it.

2) Performance issues. I understand, you need more experiments and it's slow for hyperopt, but just take a look at the implementation of the backtest() method.

If position stacking is disabled, the bot executes the `continue` statement at https://github.com/freqtrade/freqtrade/blob/bc1b5f477d8bbcb91118e639dc7445e5a9abf9e1/freqtrade/optimize/backtesting.py#L365 and goes to next iteration in the list. If position stacking is enabled, it will go further and call heavy `_get_sell_trade_entry()` method. And, remember, this will happen for every candle in a long raw of candles with buy signal...

I found long time ago that the performance of hyperopt (backtesting) evaluation depends significantly on the number of trades in the attempt. That's one of the reasons why current implementation of the hyperopt cannot utilze 100% of cpus -- short evals computed in parallel on the cpu cores simply wait for longer ones (remember, I wrote it in the PR which is posponed?). You can find it easily -- run `freqtrade hyperopt -j 1 --print-all` and you will see how performance depend on the number of trades... (evals with small amount of trades are pretty fast)

The point here is that --eps only adds trades. So it only can slow down the performance of hyperopt.


So my suggestion is to:

* Change default for position stacking to false for hyperopt, as proposed in #2009. Same for use_max_open_trades. (have not looked at it yet)
* Describe in the documentation in more details how to treat the results of hyperopt -- those consideration about masked trades you explained above.

**Comment by xmatthias** [comment]
thinking about it you're probably right, we're optimizing against something we cannot reach - so the optimization is basically worthless.

I also did a few tests (not very extensive) and the performance penalty seems to vary greatly (your version is often faster) - but it depends on the srategy more than on anything else.

Let's correct the documentation, then #2009 is good to go :+1:

**Comment by hroff-1902** [comment]
> thinking about it you're probably right, we're optimizing against something we cannot reach - so the optimization is basically worthless.

Exactly! 🥇

**Comment by sidnicholas** [comment]
@hroff-1902 
@xmatthias 

I am having the same issue mentioned in above.  I read through the conversation above but it was not clear to me how to apply the solution.  Is this something you can help me better understand?

Thank you.

**Comment by hroff-1902** [comment]
It's already merged into the develop branch. Update your freqtrade installation to current develop.

**Comment by stonetown1975** [comment]
@hroff-1902 thanks for fixing this issue. I was very confused that backtesting produces different results than hyperopt
